x/programs/rust/wasmlanche-sdk/src/context.rs

```diff
@@ -0,0 +1,129 @@
+use crate::{
+    state::{Error, Key, State},
+    types::Address,
+    Gas, Id, Program,
+};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::{cell::RefCell, collections::HashMap};
+
+/// Representation of the context that is passed to programs at runtime.
+#[cfg_attr(feature = "debug", derive(Debug))]
+pub struct Context<K = ()> {
+    program: Program,
+    actor: Address,
+    height: u64,
+    timestamp: u64,
+    action_id: Id,
+    state_cache: RefCell<HashMap<K, Option<Vec<u8>>>>,
+}
+
+impl<K> BorshDeserialize for Context<K> {
+    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
+        let program = Program::deserialize_reader(reader)?;
+        let actor = Address::deserialize_reader(reader)?;
+        let height = u64::deserialize_reader(reader)?;
+        let timestamp = u64::deserialize_reader(reader)?;
+        let action_id = Id::deserialize_reader(reader)?;
+
+        Ok(Self {
+            program,
+            actor,
+            height,
+            timestamp,
+            action_id,
+            state_cache: RefCell::new(HashMap::new()),
+        })
+    }
+}
+
+impl<K> Context<K> {
+    pub fn program(&self) -> &Program {
+        &self.program
+    }
+
+    pub fn actor(&self) -> Address {
+        self.actor
+    }
+
+    pub fn height(&self) -> u64 {
+        self.height
+    }
+
+    pub fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+
+    pub fn action_id(&self) -> Id {
+        self.action_id
+    }
+}
+
+impl<K: Key> Context<K> {
+    /// See [`State::get`].
+    /// # Errors
+    /// See [`State::get`].
+    pub fn get<V: BorshDeserialize>(&self, key: K) -> Result<Option<V>, crate::state::Error> {
+        State::new(&self.state_cache).get(key)
+    }
+
+    /// See [`State::store_by_key`].
+    /// # Errors
+    /// See [`State::store_by_key`].
+    pub fn store_by_key<V>(&self, key: K, value: &V) -> Result<(), crate::state::Error>
+    where
+        V: BorshSerialize,
+    {
+        State::new(&self.state_cache).store_by_key(key, value)
```

Comment:
since `flush` will get called once this `state` is dropped, we could move the cache flushing to be a method of context

https://github.com/ava-labs/hypersdk/blob/97a86924856b9dec1210c4a7faf247dae8730b3a/x/programs/rust/wasmlanche-sdk/src/state.rs#L145

Reply:
Yeah, this is my next change